### ISSUE TITLE
Concurrent compile error when using gulp

### ISSUE THREAD

**Issue by ducklord** [ChadKillingsworth/closure-compiler-npm#9]
When compiling two different outputs at the same time, I get some interesting errors. The errors printed by the compiler does not make sense. The lines that is reported by the compiler, is a merge of different parts of different files, also the output changes from each compile. When compiling each output on its own there are no errors. The setup is something like this:
```js
...
gulp.task('project1', ['deps'], function() {
    return gulp.src(['src/**/*.js', ...], {base: './'})
        .pipe(closureCompiler(compilerArgs1))
        .pipe(gulp.dest('dest/'));
});
gulp.task('project2', ['deps'], function() {
    return gulp.src(['src/**/*.js', ...], {base: './'})
        .pipe(closureCompiler(compilerArgs2))
        .pipe(gulp.dest('dest/'));
});
gulp.task('default', ['project1', 'project2']);
```

Example output:
```
/node_modules/jsbuild/node_modules/google-closure-library/closure/goog/labs/i18n/listsymbolsext.js:4706: ERROR - Parse error. Semi-colon expected
goog.labs.i18n.ListFormatSymbols_kiy with its upper left corner
                                     ^

/third-party/JST/src/canvas/layer.js:16: ERROR - Parse error. Unterminated string literal
goog.require('jst.core.ArrayC goog.testing.events.fireMouseMoveEvent(
             ^

2 error(s), 0 warning(s)
```

Is suspecting the chunking when parsing the JSON to the compile process, but the scopes appears to be correct. Any hints and clues to find the issue would also be appreciated. :)

**Comment by ChadKillingsworth** [comment]
Thanks for this. I'm guessing the way the error stream stores data for output later is interleaving between processes. It's not something I considered. I'll see if I can get it fixed.

**Comment by ducklord** [comment]
It was rather easy to reproduce in a simple test case, so here you go. :) Just `npm install` and run `gulp`
[concurrent-gulp.tar.gz](https://github.com/ChadKillingsworth/closure-compiler-npm/files/140324/concurrent-gulp.tar.gz)

**Comment by duzun** [comment]
@ducklord I'm facing the same issue, but only on OS X and Debian, not on Window.

I did try your `concurent-gulp.tar.gz` on Windows and OS X - it doesn't fail on Windows.

I hope this helps somehow.

**Comment by ChadKillingsworth** [comment]
I've recreated the issue. I hope to have time to fix it this weekend. It's also hitting my builds now too.

**Comment by ducklord** [comment]
Thanks, this fixes the simple use case. Unfortunately my real life case, persists. Have messed around in the code, but only found som minor issues, that did not fix the problem. Not sure if you would like the minor fixes? Would like to give you a fix for the concurrency problem. I will keep trying to find the error and report if i find a proper fix.

**Comment by ChadKillingsworth** [comment]
I can run three simultaneous compiles without problems - that wasn't working before.

**Comment by ducklord** [comment]
Have tried som different things, even when i append the chunked output to separate files for each compile process, i can not find the strings that the compiler reports as being wrong. Right now my best guess is actually that the error somehow is inside the compiler. Each of the output files are around 25MB so they are not that easy to compare, and find a single line of error. :)

**Comment by ChadKillingsworth** [comment]
Are you using gulp-sourcemaps? Try disabling that plugin. Input source maps are used by the compiler for error messages.

**Comment by ducklord** [comment]
I am not using gulp-sourcemaps. But i do have the "create_source_map" flag enabled. Tried disabling that as well, but did not make a difference.

**Comment by ducklord** [comment]
Observation: Works like a charm on windows. :)

**Comment by ducklord** [comment]
This is my test of the chuncking, and the output is identical. So i am pretty sure that the error is somewhere in the compiler. You can close this issue if you like, not sure there is more that can be done in the gulp plugin.

[check_chunk_output.txt](https://github.com/ChadKillingsworth/closure-compiler-npm/files/163380/check_chunk_output.txt)

**Comment by ChadKillingsworth** [comment]
@ducklord Did #11 fix the issue for you?

**Comment by duzun** [comment]
It didn't fix the issue for my case on Debian 8, using v20160208.5.0

**Comment by ducklord** [comment]
No, it did not. But could not hurt to have the proper scope. :) Using the patch i postet, i can see that the input to the process is correct, or at least the same for all executions of the process. So the bug must be deeper, and next after the chunking it is just the java process. So i am pretty sure, that the bug is in the closure compiler it self.

**Comment by ChadKillingsworth** [comment]
I'm closing this issue then.  @duzun If you can provide a repo case I'll take a look, but I use this on MacOS and don't see any problems right now.

**Comment by duzun** [comment]
Unfortunately I can't provide sources, but here is a build script that when run on my project on Debian throws an error:

`gulpfile.js`:
```js
var gulp            = require('gulp');
var $foreach        = require('gulp-foreach');
var $rename         = require('gulp-rename');
var $babel          = require('gulp-babel');
var closureCompiler = require('google-closure-compiler').gulp();

gulp.task('default', function () {
    var src = 'src/';
    return gulp.src([
        src+'*.js',
        src+'lib/*.js',
        src+'lib/api/*.js',
        src+'api/*.js',
    ], { base: src })
    .pipe($babel(babelOptions()))
    .pipe($foreach((stream, file) => {
        var relative = file.relative;
        console.log(relative);
        return stream
            .pipe(closureCompiler(compileOptions()))
            .pipe($rename(relative))
    }))
    .pipe(gulp.dest('dist/'))
    ;
});

function compileOptions(stream, options, concatPath) {
    return {
          "warning_level"     : "DEFAULT",
          "compilation_level" : "SIMPLE_OPTIMIZATIONS",
          "language_in"       : "ECMASCRIPT6",
          "language_out"      : "ECMASCRIPT3"
    }
}

function babelOptions() {
    return {
      "plugins": [
        "check-es2015-constants",
        "transform-es2015-arrow-functions",
        "transform-es2015-block-scoped-functions",
        "transform-es2015-block-scoping",
        "transform-es2015-classes",
        "transform-es2015-computed-properties",
        "transform-es2015-destructuring",
        "transform-es2015-for-of",
        "transform-es2015-function-name",
        "transform-es2015-literals",
        "transform-es2015-object-super",
        "transform-es2015-parameters",
        "transform-es2015-shorthand-properties",
        "transform-es2015-spread",
        "transform-es2015-template-literals",
        [
          "syntax-async-functions"
        ],
        [
          "transform-regenerator"
        ]
      ],
      "compact": false
    }
}
```

And here is the kind of error I'm getting:

```
events.js:154
      throw er; // Unhandled 'error' event
      ^
Error: Compilation error:

java -jar /var/project/node_modules/google-closure-compiler/compiler.jar --warning_level=DEFAULT --compilation_level=SIMPLE_OPTIMIZATIONS --language_in=ECMASCRIPT6 --language_out=ECMASCRIPT3 --json_streams BOTH

/src/file.js:2706: ERROR - Parse error. Character '�' (U+FFFD) is not a valid identifier start char
        result = action.func.apply(actii��
```

It is obvious to me there is some overlapping going on...
The error is different with each run, and sometimes it doesn't throw the error.

If I remove `.pipe($babel(babelOptions()))` or move it inside `$foreach` just right before `.pipe(closureCompiler(compileOptions()))`, I don't see the error any more.

I'm using `gulp-foreach` and `gulp-rename` just to be able to compile JS files one by one, without combining them into one big JS file.

My original `gulpfile.js` is a little bit more complex and it triggers the error with every run.
It is tough to reproduce it without my original sources (there ave over 50 js files to be compiled), but I'm sure there is a bug somewhere ...

`package.json`

```json
{
  "name": "concurrent-gulp-foreach",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "gulp"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "google-closure-compiler": "^20160208.x",
    "gulp": "^3.9",
    "gulp-foreach": "^0.1",
    "gulp-rename": "^1.2",
    "gulp-babel": "^6.x",
    "babel-plugin-transform-async-to-module-method": "^6.4",
    "babel-preset-es2015": "^6.3",
    "babel-preset-stage-3": "^6.3"
  }
}
```

**Comment by ChadKillingsworth** [comment]
I'll setup a test case locally with a gulp-foreach call.

**Comment by ducklord** [comment]
Did a test in the closure compiler. Wrote the input stream to a file. The result was each of the processies wrote a file of exactly the same size, given the same input. *But* the content was out of order, and I could find the malformed content that the compiler complains about in only one of the files. So my assumption that the problem was in the compiler was wrong, did not make any sense any way. Could be a NodeJS issue? That only appears on Linux? That stream.write() does not guarantee in order writing when running more then one child_process.spawn()?

**Comment by ChadKillingsworth** [comment]
> That stream.write() does not guarantee in order writing when running more then one child_process.spawn()?

The source files are all gathered into one huge JSON object and that single object is written to the compiler. The only place I could see this affecting things is the back pressure logic for writing to stdin. If the JSON object is out of order, that would be a different issue.

Try writing the file paths from `this.filelist_` to a file here: https://github.com/ChadKillingsworth/closure-compiler-npm/blob/master/lib/gulp/index.js#L90  That will at least ensure that the files were actually passed to the compiler in the order you expect. Something like:

```js
let outputFileList = [];
for (let index = 0; index < this.filelist_.length; index++) {
  outputFileList.push(this.filelist_[index].path);
}
fs.writeFileSync(randomFileName, outputFileList.join('\n'));
```

If that's correct, do the same thing with the json stream results just before it is written to stdin. This will help narrow down where the issue is occurring.

**Comment by ducklord** [comment]
As I wrote earlier, I changed "write_buffer_in_chunks()" a bit, so that it also write the chunks to a file using fs.appendFileSync(). The output in the files are valid, and correct. But the java processes still receives the chunks out of order.

**Comment by ChadKillingsworth** [comment]
@ducklord I didn't catch that you had tested this within the `write_buffer_in_chunks` method.

Can you try the changes in #12 and see if that alleviates the issue?

**Comment by ChadKillingsworth** [comment]
I had travis configured wrong, so they are in master now. If those changes cause problems, I'll roll them back.

**Comment by ducklord** [comment]
That fixes the issue! Thank you very very much. Can still not see why your old chunking method did not work. But readable stream does the trick. :)

**Comment by Dominator008** [comment]
Thanks @ducklord. @duzun would you like to test against the current master?

**Comment by ChadKillingsworth** [comment]
Closed by #12

**Comment by duzun** [comment]
I just run my build script on Debian 8 and no issue!
Looks like it is fixed, thanks :-)

**Comment by rsimha** [comment]
@ChadKillingsworth We're seeing this issue with the latest versions of closure compiler while compiling the [AMP project](https://amp.dev).

Specifically, concurrent invocations of closure compiler while using compiler flags (like `--pseudo_names` or `--source_map_include_content`) that tend to make compilation slower are most likely to yield a JSON parse error.

Let me know if you'd rather I opened a new issue, or if you'd like me to post detailed repro steps.
